homeassistant/components/calendar/__init__.py

```diff
@@ -0,0 +1,265 @@
+    'start': date || dateTime as str,
+    'end' date || dateTime as str,
+}  # can contain more but won't be used
+or
+None # if no events are found
+"""
+import logging
+import re
+
+from datetime import timedelta
+from homeassistant.util import dt
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.entity import Entity
+from homeassistant.const import (STATE_ON, STATE_OFF)
+from homeassistant.util import Throttle
+from homeassistant.components import (
+    google)
+
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'calendar'
+ENTITY_ID_FORMAT = DOMAIN + '.{}'
+
+DISCOVERY_PLATFORMS = {
+    google.DISCOVER_CALENDARS: 'google',
+}
+
+
+def setup(hass, config):
+    """Track states and offer events for calendars."""
+    component = EntityComponent(
+        logging.getLogger(__name__), DOMAIN, hass, 60, DISCOVERY_PLATFORMS)
+
+    component.setup(config)
+
+    return True
+
+DEFAULT_CONF_TRACK_NEW = True
+DEFAULT_CONF_OFFSET = '#-'
+
+# Return cached results if last scan was less then this time ago
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=5)
+
+
+# pylint: disable=too-many-instance-attributes
+class CalendarEventDevice(Entity):
+    """A calendar event device."""
+
+    # pylint: disable=too-many-arguments
+    def __init__(self, hass, calendar, data):
+        """Create the Calendar Event Device.
+
+        hass = HA instance
+        calendar = reference for subclasses to get main cal object
+        data = dict of track/search/name/offset
+        """
+        from homeassistant.helpers.entity import generate_entity_id
+        self.hass = hass
+        self._state = False
+        self._calendar = calendar
+        self._track = data.get('track', DEFAULT_CONF_TRACK_NEW)
+        self._search = data.get('search', None)
+        self._name = data.get('name')
+        self._offset = data.get('offset', DEFAULT_CONF_OFFSET)
+        self.entity_id = generate_entity_id(ENTITY_ID_FORMAT,
+                                            self._name,
+                                            hass=hass)
+        self._start_listener = None
+        self._end_listener = None
+
+        self._cal_data = {
+            'all_day': False,
+            'offset_time': 0,
+            'message': '',
+            'start': None,
+            'end': None,
+            'location': None,
+            'description': '',
+        }
+
+        if self._track:
+            self.update()
+
+    @property
+    def offset_reached(self):
+        """Have we reached the offset time specified in the event title."""
+        if self._cal_data['offset_time'] > 0:
+            start = self._cal_data['start']
+            now = dt.now(start.tzinfo)
+            offset = timedelta(minutes=self._cal_data['offset_time'])
+            return now + offset >= start
+        return False
+
+    @property
+    def state_attributes(self):
+        """State Attributes for HA."""
+        start = self._cal_data.get('start', None)
```

Comment:
Brain fart, supposed to return those, fixed. Had to copy from my prod system since I couldn't replicate in my dev.